e2etest/layer2_test.go

```diff
@@ -275,13 +275,102 @@ var _ = ginkgo.Describe("L2", func() {
 	},
 		table.Entry("IPV4", &ipv4ServiceRange),
 		table.Entry("IPV6", &ipv6ServiceRange))
+
+	table.DescribeTable("validate requesting a specific address pool for Loadbalancer service", func(ipRange *string) {
+		var services [2]*corev1.Service
+		var servicesIngressIP [2]string
+		var pools []addressPool
+
+		namspace := f.Namespace.Name
+
+		for i := 0; i < 2; i++ {
+			ginkgo.By(fmt.Sprintf("configure addresspool number %d", i+1))
+			ip := getIPFromRangeByIndex(*ipRange, i)
+			addressesRange := fmt.Sprintf("%s-%s", ip, ip)
+			pool := addressPool{
+				Name:     fmt.Sprintf("test-addresspool%d", i+1),
+				Protocol: Layer2,
+				Addresses: []string{
+					addressesRange,
+				},
+			}
+			pools = append(pools, pool)
+
+			configData := configFile{
+				Pools: pools,
+			}
+			err := updateConfigMap(cs, configData)
+			framework.ExpectNoError(err)
+
+			ginkgo.By(fmt.Sprintf("configure service number %d", i+1))
+			jig := e2eservice.NewTestJig(cs, namspace, fmt.Sprintf("test-service%d", i+1))
+			timeout := e2eservice.GetServiceLoadBalancerCreationTimeout(cs)
+			svc, err := jig.CreateLoadBalancerService(timeout, func(svc *corev1.Service) {
+				tweakServicePort(svc)
+				svc.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+				svc.Annotations = map[string]string{"metallb.universe.tf/address-pool": fmt.Sprintf("test-addresspool%d", i+1)}
+			})
+			framework.ExpectNoError(err)
+
+			_, err = jig.Run(func(rc *corev1.ReplicationController) {
+				tweakRCPort(rc)
+			})
+			framework.ExpectNoError(err)
+
+			defer func() {
+				err := cs.CoreV1().Services(svc.Namespace).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err)
+			}()
+
+			ginkgo.By("validate LoadBalancer IP is in the AddressPool range")
+			ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+			err = validateIPInRange([]addressPool{pool}, ingressIP)
+			framework.ExpectNoError(err)
+
+			services[i] = svc
+			servicesIngressIP[i] = ingressIP
+
+			for j := 0; j <= i; j++ {
+				ginkgo.By(fmt.Sprintf("validate service %d IP didn't change", j+1))
+				ip := e2eservice.GetIngressPoint(&services[j].Status.LoadBalancer.Ingress[0])
+				framework.ExpectEqual(ip, servicesIngressIP[j])
+
+				ginkgo.By(fmt.Sprintf("checking connectivity of service %d to its external VIP", j+1))
+				port := strconv.Itoa(int(services[j].Spec.Ports[0].Port))
+				hostport := net.JoinHostPort(ip, port)
+				address := fmt.Sprintf("http://%s/", hostport)
+				err = wgetRetry(address, executor.Host)
+				framework.ExpectNoError(err)
+			}
+		}
+	},
+		table.Entry("IPV4", &ipv4ServiceRange),
+		table.Entry("IPV6", &ipv6ServiceRange))
 })
 
-func firstIPFromRange(ipRange string) string {
-	cidr, err := internalconfig.ParseCIDR(ipRange)
+func getIPFromRangeByIndex(ipRange string, index int) string {
+	cidrs, err := internalconfig.ParseCIDR(ipRange)
```

Comment:
why is not enough to pick the first cidr anymore?
IIUC. by doing this we go over all the possible cidrs related to the range, which is different from picking the first and the nth item of the same cidr.

Reply:
for example, when passing the following range `172.18.0.6-172.18.0.16` to the ParseCIDR function we get the following cidrs:`[172.18.0.6/31 172.18.0.8/29 172.18.0.16/32]`
if we want to get the 5th ip in the range we can't use the first cidr as it has only two ips.

Reply:
so I think a better option is to pick the largest CIDR and use that as a range reference. Otherwise we (at extreme) risk picking the same ip because you jump from one CIDR to another

Reply:
but the cidrs are not overlapping so i can't get the same ip, or they might me overlapping?

Reply:
ugh, I completely mis-understood what ParseCIDR does.Ignore what I just said, thanks for explaining.